File: frontend/src/validators/userValidator.ts
export interface NewUser {
    username: string;
    firstname: string;
    lastname: string;
    gender: string;
    password: string;
    score?: number;
    role?: string;    
  }
  
export const validateUserInput = (user: NewUser): string | null => {
  const allowedGenders = ["male", "female", "other"];
  const { username, firstname, lastname, gender, password } = user;

  if (!username) return "Username is required.";
  if (username.trim().length < 3 || username.length > 80) return "Username must be 3–80 characters.";

  if (!firstname) return "First name is required.";
  if (firstname.length > 80) return "First name must be ≤ 80 characters.";

  if (!lastname) return "Last name is required.";
  if (lastname.length > 80) return "Last name must be ≤ 80 characters.";

  if (!gender) return "Gender is required.";
  if (!allowedGenders.includes(gender.toLowerCase())) return "Gender must be male, female, or other.";

  if (!password) return "Password is required.";
  if (password.length < 5) return "Password must be at least 5 characters.";

  return null;
};

File: backend/app/validators/user_validator.py
from marshmallow import Schema, fields, validate, ValidationError


class CreateUserSchema(Schema):
    username = fields.Str(required=True, validate=validate.Length(min=3, max=30))
    password = fields.Str(required=True, validate=validate.Length(min=6))
    firstname = fields.Str(required=True, validate=validate.Length(min=1))
    lastname = fields.Str(required=True, validate=validate.Length(min=1))
    gender = fields.Str(required=True, validate=validate.OneOf(["male", "female", "other"]))
    score = fields.Int(missing=0)
    role = fields.Str(missing="user", validate=validate.OneOf(["user", "admin"]))


class UpdateUserSchema(Schema):
    firstname = fields.Str(validate=validate.Length(min=1))
    lastname = fields.Str(validate=validate.Length(min=1))
    gender = fields.Str(validate=validate.OneOf(["male", "female", "other"]))
    score = fields.Int()
    role = fields.Str(validate=validate.OneOf(["user", "admin"]))
    password = fields.Str(validate=validate.Length(min=6))


def validate_create_user_payload(payload: dict) -> dict:
    """
    Validates and transforms data for user creation.
    Raises ValidationError if data is invalid.
    """
    schema = CreateUserSchema()
    result = schema.load(payload)
    return result


def validate_update_user_payload(payload: dict) -> dict:
    """
    Validates and transforms data for user update.
    Ignores unknown fields. Raises ValidationError if invalid.
    """
    schema = UpdateUserSchema()
    result = schema.load(payload, partial=True)
    return result

def validate_delete_user_payload(user) -> None:
    """
    Prevents deletion of the original super admin account.
    """
    if user.username == "admin" and user.role == "admin":
        raise ValidationError("The original super admin cannot be deleted.")